Wabamedia: WC Plugin Update

wabamonsta opened the pull request:
### Customer Management
- Update Transaction  to show Order # and Product(s) Name in the Transaction Description
-  Add order # to the Transaction 
- Set Digital product to complete once Payload Authorize the transaction 
- Check  if Customer exist on payload before we create them 

### Payment Token Handling
- Default User Token  to the Default Payment CC
-  Check if CC already exist in order to prevent generating a new one 

### Error Handling & Notification 
- Modify Unit Test (WIP)

## What's Changed
* Update transaction description to show order and porducts by @wabamonsta in https://github.com/Wabamedia/payload-woocommerce/pull/2
* add lazy load to index.jsx function also double check if form is alre… by @wabamonsta in https://github.com/Wabamedia/payload-woocommerce/pull/3
* Ensure wc status update is reflected on payload processed by @wabamonsta in https://github.com/Wabamedia/payload-woocommerce/pull/6
* Customermanagement Enhancement by @wabamonsta in https://github.com/Wabamedia/payload-woocommerce/pull/7
* Modify Unit test by @wabamonsta in https://github.com/Wabamedia/payload-woocommerce/pull/10
* Use user default token default payment cc by @wabamonsta in https://github.com/Wabamedia/payload-woocommerce/pull/11
* Payment token handling by @wabamonsta in https://github.com/Wabamedia/payload-woocommerce/pull/12
* Merge pull request #10 from Wabamedia/jermaine-wab-29-modify-unit-test by @wabamonsta in https://github.com/Wabamedia/payload-woocommerce/pull/13
* Credit card form loading multiple times then timeout by @wabamonsta in https://github.com/Wabamedia/payload-woocommerce/pull/18
* Add order number to payload receipt by @wabamonsta in https://github.com/Wabamedia/payload-woocommerce/pull/19


**Full Changelog**: https://github.com/Wabamedia/payload-woocommerce/commits/1.3.0-rc2

wabamonsta committed:
allow wc to check if wc_subscription function exist

wabamonsta committed:
check if customer exist on payload before creating a new one

wabamonsta committed:
- Add order description to payload transaction
- Add utility function to capture product name and status to detect if
  the items purchase is a virtual item

wabamonsta committed:
Add description to transaction

wabamonsta committed:
updated payment description

wabamonsta committed:
remove test code

wabamonsta committed:
WAB-19: cleared space also testing linear sync

wabamonsta committed:
Add git workflow to auto fill PR Description with previous commits

wabamonsta committed:
update github workflow

wabamonsta committed:
Merge pull request #2 from Wabamedia/jermaine/wab-19-update-transaction-description-to-show-order-and-porducts

Jermaine/wab 19 update transaction description to show order and porducts

wabamonsta committed:
remove extra workflow file

wabamonsta committed:
Add additional condition for when admin is updating customer payment method

wabamonsta committed:
- Auto complete payment if it is a virtual item and the payment is successful
- Add description to all payment
- Query payload customer by email upon creation to prevent duplicate
- When procesing payment user id is retrieve from Order for additional
  safety instead of relying on current user id

wabamonsta committed:
add lazy load to index.jsx function also double check if form is already mounted

wabamonsta committed:
Merge pull request #3 from Wabamedia/jermaine-wab-21-fix-jsx-credit-card-fields-ensure-its-loaded-properly

add lazy load to index.jsx function also double check if form is alre…

wabamonsta committed:
update Auto fill Pr description workflow

wabamonsta committed:
Merge pull request #6 from Wabamedia/jermaine-wab-20-ensure-wc-status-update-is-reflected-on-payload-processed

Jermaine wab 20 ensure wc status update is reflected on payload processed

wabamonsta committed:
update PR workflow

wabamonsta committed:
create helper function to resolve phpunit test issue

wabamonsta committed:
Fix jsx lint issue

wabamonsta committed:
update gitignore

wabamonsta committed:
update gitignore

wabamonsta committed:
update unit test

wabamonsta committed:
update unit test

wabamonsta committed:
update unit test add gaurd to check for Customer::filter_by

wabamonsta committed:
Merge pull request #7 from Wabamedia/jermaine/customermanagement

Jermaine/customermanagement

wabamonsta committed:
update gitignore

wabamonsta committed:
remove tracking .phpunit.result.cache

wabamonsta committed:
- Create Customer Filtered by Mock
- Create Customer First function

wabamonsta committed:
Merge pull request #10 from Wabamedia/jermaine-wab-29-modify-unit-test

WAB-29  Modify Unit test

wabamonsta committed:
check if user card already exist and if it does then app use the token id assigned to the original card

wabamonsta committed:
Check for Payload customerID with email and if it it existthen assign it as payload customer id

wabamonsta committed:
add return to filter

wabamonsta committed:
Merge pull request #11 from Wabamedia/jermaine-wab-27-use-user-default-token-default-payment-cc

wab 27 use user default token default payment cc

wabamonsta committed:
Update Mock with intention to merge

wabamonsta committed:
Merge pull request #12 from Wabamedia/jermaine-wab-26-payment-token-handling

Jermaine wab 26 payment token handling

wabamonsta committed:
Merge pull request #13 from Wabamedia/jermaine-wab-28-error-handling-notifications

Merge pull request #10 from Wabamedia/jermaine-wab-29-modify-unit-test

wabamonsta committed:
add order number to payload receipt

wabamonsta committed:
fix creditcard form load  issuse on checkout

wabamonsta committed:
fix cc form loading issue on checkout and user add payment method

wabamonsta committed:
Fix lint issue

wabamonsta committed:
Merge pull request #18 from Wabamedia/jermaine-wab-33-credit-card-form-loading-multiple-times-then-timeout

Jermaine wab 33 credit card form loading multiple times then timeout

wabamonsta committed:
Merge pull request #19 from Wabamedia/jermaine-wab-30-add-order-to-payload-transaction

add order number to payload receipt

wabamonsta committed:
remove extra component lock file

wabamonsta committed:
add get_items to wc mock

wabamonsta committed:
add release notes workflow

wabamonsta committed:
update version id

wabamonsta committed:
update release workflow

wabamonsta committed:
update workflow

wabamonsta committed:
update workflow wip

wabamonsta committed:
update workflow wip

wabamonsta committed:
update work flow

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -28,9 +28,11 @@ public function __construct() {
 		add_action( 'woocommerce_update_options_payment_gateways_' . $this->id, array( $this, 'process_admin_options' ) );
 		add_action( 'woocommerce_scheduled_subscription_payment_' . $this->id, array( $this, 'scheduled_subscription_payment' ), 10, 2 );
 		add_action( 'wp_enqueue_scripts', array( $this, 'payment_scripts' ) );
+//		$this->__testpayloadgateway();
```
Lets remove this commented out line.

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -256,36 +269,157 @@ public function scheduled_subscription_payment( $amount, $renewal_order, $retry
-				'description'       => 'Payment for order #' . $order->get_id(),
+		$order_id =  $order->get_id();
+		$payment_array = array(
+				'description'       =>  'Payment for order #' . $order_id." related to  Product: ".$this->get_order_product_name($order_id),
```
Wondering if we can get rid of `'Payment for order #' . $order_id." ` since the $order_id is set as the `order_number` below on line 277.

wabamonsta replied:
I Agree

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -171,15 +179,20 @@ public function process_payment( $order_id ) {
-		$order->payment_complete();
-		$order->save();
+	
+		if($payment->status  == 'authorized' ){
```
I think we should check if it's a virtual payment here, and only set the status to processed if it's authorized and not processed.

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -171,15 +179,20 @@ public function process_payment( $order_id ) {
+			$payment->update( array('order_number'=>strval( $order_id),  'status' => 'processed', "description"=> 'Payment for order #' . $order_id." related to  Product: ".$this->get_order_product_name($order_id) ) );
+	
+				$order->payment_complete();
+				$order->save();
```
I think this should be moved out of this conditional, because sometimes the payment status will already be set to processed, and in that case we should set the payment to complete.

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -256,36 +269,157 @@ public function scheduled_subscription_payment( $amount, $renewal_order, $retry
 		$order->set_transaction_id( $payment->ref_number );
-		$order->payment_complete();
-
+	//	if( $this->is_virtual($order_id)){
```
Should this be uncommented? Wondering if this should be moved to a function called something like `handle_order_payment` that we can reuse in `create_payment_for_order` and `process_payment`. So something like:

```php
public function handle_order_payment( $order, $payment ) {
    $order->set_transaction_id( $payment->ref_number );
    if($payment->status  == 'authorized' ){
        $payment->update( array('order_number'=>strval( $order_id),  'status' => 'processed', "description"=> $this->get_order_product_name($order_id) ) );
    }
    if ($payment->status  == 'processed' && $this->is_virtual($order_id)){
			$order->payment_complete();
			$order->save();
    }
}
```

wabamonsta replied:
I agree

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -256,36 +269,157 @@ public function scheduled_subscription_payment( $amount, $renewal_order, $retry
+		$order = wc_get_order( $order_id );
+		$items = $order->get_items();
+
+		foreach ( $items as $item ) {
+			if ( $item->get_product()->is_virtual() ) {
+				return true;
+			}
+		}
```
Should this only return true only if all are virtual? So invert this to:

```php
       foreach ( $items as $item ) {
			if ( !$item->get_product()->is_virtual() ) {
				return false;
			}
		}
        return true;
```

wabamonsta replied:
Yes that make a lot of sense. I agree

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -256,36 +269,157 @@ public function scheduled_subscription_payment( $amount, $renewal_order, $retry
+
+
+
+	public function set_customer_id_by_order($order){
```
I'm not sure I understand what this function is solving.

wabamonsta replied:
When admin accessing customer Order the application use the Customer Id on the order instead of using WP current user id. So in the case where administrator attempt to manage a customer account remotely it doesnt consider using the Admin ID on the Order/Transaction update

ianhalpern commented on src/js/index.jsx:
```diff
@@ -248,20 +249,116 @@ const mountPaymentMethodForm = () => {
 			'#payload-add-payment-method'
 		);
-
+		//console.log('[payload] Mounting payment method form into', domContainer);
```
Lets remove this commented out line.

ianhalpern commented on src/js/index.jsx:
```diff
@@ -248,20 +249,116 @@ const mountPaymentMethodForm = () => {
+		// If we have a container or had previously found it (pending), mount now.
+		if ( container || pending ) {
+			actuallyMount(
+				container || document.querySelector( TARGET_SELECTOR )
```
What's the ` || document.querySelector( TARGET_SELECTOR )`? We've already set `const container = document.querySelector( TARGET_SELECTOR );` above.

ianhalpern commented on src/js/index.jsx:
```diff
@@ -248,20 +249,116 @@ const mountPaymentMethodForm = () => {
+		// Normalize: if caller passed `true` before, just re-query the DOM
+		if ( ! container || container.nodeType !== 1 ) {
+			container = document.querySelector( TARGET_SELECTOR );
+		}
```
I don't think this is necessary.

ianhalpern commented on src/js/index.jsx:
```diff
@@ -248,20 +249,116 @@ const mountPaymentMethodForm = () => {
+		const container = document.querySelector( TARGET_SELECTOR );
+
+		// If we have a container or had previously found it (pending), mount now.
+		if ( container || pending ) {
```
Not following why we'd allow this to continue if  `pending` is `true` but the container doesn't exist.

ianhalpern commented on src/js/index.jsx:
```diff
@@ -248,20 +249,116 @@ const mountPaymentMethodForm = () => {
+	const TARGET_SELECTOR = '#payload-add-payment-method';
+
+	let mountedContainer = null; // track which exact element we mounted into
+	let pending = false;
```
I'm not sure we need this.

wabamonsta replied:
which line is in question here ?

ianhalpern replied:
I was referring to `let pending = false;`

ianhalpern commented on payload-woocommerce.php:
```diff
@@ -65,30 +65,74 @@ function ( Automattic\WooCommerce\Blocks\Payments\PaymentMethodRegistry $payment
 function get_payload_customer_id() {
 	$payload_customer_id = null;
 
 	$user = wp_get_current_user();
```
Wondering, based on your other changes, should we be only trying to grab the user from an order?

wabamonsta replied:
You have to consider what if there is no Order, could be a case where the user is just adding a new Payment method so no order to associate themself with.  The scenario where I use the `set_customer_id_by_order` there is an active order

ianhalpern reviewed (changes requested):
Left some comments, also looks like there's some formatting that needs to be cleaned up. Wonder if we should add a linting step in CI to run:

`vendor/bin/phpcs --standard=WordPress`

and

`npm run lint:css lint:js lint:md:docs lint:md:js lint:pkg-json`

wabamonsta committed:
check log

wabamonsta committed:
update base on change request

wabamonsta commented on includes/class-wc-payload-gateway.php:
```diff
@@ -32,7 +32,6 @@ public function __construct() {
 		add_action( 'woocommerce_update_options_payment_gateways_' . $this->id, array( $this, 'process_admin_options' ) );
 		add_action( 'woocommerce_scheduled_subscription_payment_' . $this->id, array( $this, 'scheduled_subscription_payment' ), 10, 2 );
 		add_action( 'wp_enqueue_scripts', array( $this, 'payment_scripts' ) );
-//		$this->__testpayloadgateway();
```
Removed as per requested

wabamonsta commented on includes/class-wc-payload-gateway.php:
```diff
@@ -281,17 +280,24 @@ public function create_payment_for_order( $order, $amount, $payment_method_id )
+			$order->payment_complete();
+			
+    }
+	$order->save();
```
I move the  `$order->save()` outside the condition so it execute for non virtual items as well

wabamonsta committed:
when customer information is updated on the back office the information is sync with Payload if they're an active customer on Payload

wabamonsta committed:
update comments

wabamonsta committed:
fix linting issue and update test workflow

wabamonsta committed:
update composer and vscode rules

wabamonsta committed:
update package.json

wabamonsta committed:
update package json to fix lint error on deployment

wabamonsta committed:
update package json to fix lint error on deployment

wabamonsta committed:
update package json to fix lint error on deployment

wabamonsta committed:
update package json to fix lint error on deployment reorder package.json file

wabamonsta committed:
update package json to fix lint error on deployment reorder package.json file

wabamonsta committed:
Merge pull request #22 from Wabamedia/jermaine-wab-31-add-company-name-as-part-of-the-customer-name-on-payload

add company name as part of the customer name on payload

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -88,24 +89,27 @@ public function process_payment( $order_id ) {
 
-		// Update subscription payment method
-		if ( wcs_is_subscription( $order_id ) ) {
+		$get_user_id_from_order = $this->get_order_customer_id($order);	
```
Since this is a variable, thinking we should rename to `$user_id_from_order`

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -171,15 +178,15 @@ public function process_payment( $order_id ) {
-			if ( WC_Subscriptions_Order::order_contains_subscription( $order_id ) ) {
-				$token = $this->create_token( $payment->payment_method );
+			if ( function_exists("WC_Subscriptions_Order") && WC_Subscriptions_Order::order_contains_subscription( $order_id ) ) {
+				$token = $this->create_token( $payment->payment_method , $this->set_customer_id_by_order($order)  );
```
Thinking it might make more sense to update `set_customer_id_by_order` to not return the order user id, since the function is just supposed to set the customer id. So instead do this:

```php
$this->set_customer_id_by_order( $order )
$token = $this->create_token( $payment->payment_method ,   $get_user_id_from_order );
```

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -256,36 +263,164 @@ public function scheduled_subscription_payment( $amount, $renewal_order, $retry
-				'description'       => 'Payment for order #' . $order->get_id(),
+		$order_id =  $order->get_id();
+		$payment_array = array(
+				'description'       =>  " Order Item(s): ".$this->get_order_product_name($order_id),
```
Looks like there's a leading space that should be removed.

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -256,36 +263,164 @@ public function scheduled_subscription_payment( $amount, $renewal_order, $retry
+    $order_id =  $order->get_id();
+    if($payment->status  == 'authorized' ){
+        $payment->update( array('order_number'=>strval( $order_id),  'status' => 'processed', "description"=> " Order Item(s): ".$this->get_order_product_name($order_id) ) );
+        $get_user_company = get_user_meta( $order->get_user_id(), 'billing_company', true );
```
Being a variable, I think it would make more sense to remove the verb and be: `$user_company`

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -256,36 +263,164 @@ public function scheduled_subscription_payment( $amount, $renewal_order, $retry
+	public function handle_order_payment( $order, $payment ) {
+    $order->set_transaction_id( $payment->ref_number );
+    $order_id =  $order->get_id();
+    if($payment->status  == 'authorized' ){
```
I think this should be:

```php
if($payment->status  == 'authorized' && $this->is_virtual($order_id)){
```

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -256,36 +263,164 @@ public function scheduled_subscription_payment( $amount, $renewal_order, $retry
+            $payment->update( array('attrs' => array( 'Company Name' => $get_user_company ) ) );
+        }
+    }
+    if ($payment->status  == 'processed' && $this->is_virtual($order_id)){
```
I think this should just be:

```php
if ($payment->status  == 'processed') {
```

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -256,36 +263,164 @@ public function scheduled_subscription_payment( $amount, $renewal_order, $retry
+								$parent_payment_method = Payload\PaymentMethod::get( $parent_pm_id );
+								if ( ! empty( $parent_payment_method->account_id ) ) {
+									$payload_customer_id = $parent_payment_method->account_id;
+								}
```
I think this should be:

```php
								if ( ! empty( $parent_payment_method->customer_id ) ) {
									$payload_customer_id = $parent_payment_method->customer_id;
								}
```

ianhalpern reviewed (commented):
Made some more comments but changes are looking good. There are some formatting issues we should cleanup as well. Also, after looking in the wp database, I think we might want to switch from using `user_nicename` to `user_displayname`.

wabamonsta committed:
- System provide better error response instead of breaking the backoffice when subscription user dont have payment method on file
- System attempt to retry On-hold payment if customer update their card
- Admin notices add for better on screen error reporting
- Admin and Customer get email notification when payment failed they get notify as to what action should be taken next

wabamonsta committed:
Merge pull request #23 from Wabamedia/jermaine-wab-38-put-a-trigger-in-place-if-card-is-updated-and-an-order-was

Put a trigger in place if card is updated and an Order was placed on Hold in the Last 30 Days should be updated to Pending

ianhalpern commented on .github/workflows/pr-fill-from-commits.yml:
I'm not sure about keeping this in the main repo. We follow a different PR description format generally.

ianhalpern commented on includes/class-wc-payload-gateway.php:
Looks like a bug where we aren't grabbing the `$payment_method` before hitting this line.

wabamonsta replied:
I agree, this will be updated

ianhalpern commented on payload-woocommerce.php:
```diff
@@ -15,6 +15,67 @@
 require_once 'vendor/autoload.php';
 use Payload\API as pl;
+// Force a Company field to show on the billing section
+add_action( 'woocommerce_after_checkout_billing_form', function( $checkout ) {
+
+    echo '<div class="form-row form-row-wide" id="billing_company_custom_wrapper">';
+
+    woocommerce_form_field(
+        'billing_company',
+        [
+            'type'        => 'text',
+            'class'       => [ 'form-row-wide' ],
+            'label'       => __( 'Company Name', 'woocommerce' ),
+            'required'    => false,
+            'priority'    => 8,
+        ],
+        $checkout->get_value( 'billing_company' )
+    );
+
+    echo '</div>';
+}, 10, 2 );
```
Wondering if this should be a settings option, so it can be enabled or disabled if the customer wants it included or not.

wabamonsta replied:
Company field is not a Required field so if the User didnt fill it out it wont affect the transaction . So we stating it as enabled or disabled isnt really an issue

ianhalpern commented on payload-woocommerce.php:
```diff
@@ -15,6 +15,67 @@
+            $customer = Payload\Customer::filter_by( array('id'=>$payload_customer_id) )->update(
+                array(
+                    'email' => $user->user_email,
+                    'name'  => $user->user_nicename,
```
I think this should actually be `$user->display_name`

ianhalpern commented on payload-woocommerce.php:
```diff
@@ -15,6 +15,67 @@
+
+        // Also persist on the user, if logged in
+        if ( $order->get_customer_id() ) {
+            update_user_meta( $order->get_customer_id(), 'billing_company', $company );
```
I thought there was a standard field for `billing_company` on the user object. Not sure I like having to use custom metadata for this.

ianhalpern commented on payload-woocommerce.php:
```diff
@@ -15,6 +15,67 @@
+                    'name'  => $user->user_nicename,
+                    'attrs' => array(
+                        '_wp_user_id' => $user_id,
+                        'Company Name'=>get_user_meta( $user_id, 'billing_company', true
```
Instead of this being a custom field on the payload side, should this be the `name` field if set.

ianhalpern commented on payload-woocommerce.php:
```diff
@@ -65,30 +126,114 @@ function ( Automattic\WooCommerce\Blocks\Payments\PaymentMethodRegistry $payment
+
+							// Create new Payload customer
+						$customer = Payload\Customer::create(
+							array(
+								'email' => $user->user_email,
+								'name'  => $user->user_nicename,
+								'attrs' => array(
+									'_wp_user_id' => $user->ID,
+									'Company Name'=>get_user_meta( $user->ID, 'billing_company', true ),
+								),
+							)
+						);
```
I think it would make sense to create a `build_paylaod_customer_request` that can be reused with `profile_update` function

ianhalpern commented on includes/class-wc-payload-gateway.php:
```diff
@@ -256,36 +307,239 @@ public function scheduled_subscription_payment( $amount, $renewal_order, $retry
+        $payment->update( array('order_number'=>strval( $order_id),  'status' => 'processed', "description"=> " Order Item(s): ".$this->get_order_product_name($order_id) ) );
+        $get_user_company = get_user_meta( $order->get_user_id(), 'billing_company', true );
+        if(!empty($get_user_company)){
+            $payment->update( array('attrs' => array( 'Company Name' => $get_user_company ) ) );
```
We already have an update call on line 330. We should just include this in that request when applicable.

ianhalpern commented on payload-woocommerce.php:
```diff
@@ -149,6 +293,120 @@ function setup_payload_api() {
+}
+
+function payload_card_update_retry_suppressed( $status = null ) {
+	static $suppressed = false;
```
This could cause a race condition with concurrent requests. Maybe make this flag order specific.
